### PR TITLE
K8SPS-307 fix smart update

### PR DESCRIPTION
**Pull request by nmarukovich**
[K8SPS-307](https://jira.percona.com/browse/K8SPS-307) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

**CHANGE DESCRIPTION**
---
**Problem:**
*Short explanation of the problem.*
The logs about successful smart update were missed. 
**Cause:**
*Short explanation of the root cause of the issue if applicable.*
We compare revision hash from old container with new sts hash. I added getting of a new revision hash and now we compare sts hash with primary pod new revision hash. 
**Solution:**
*Short explanation of the solution we are providing with this PR.*

**CHECKLIST**
---
**Jira**
- [ ] Is the Jira ticket created and referenced properly?
- [ ] Does the Jira ticket have the proper statuses for documentation (`Needs Doc`) and QA (`Needs QA`)?
- [ ] Does the Jira ticket link to the proper milestone (Fix Version field)?

**Tests**
- [ ] Is an E2E test/test case added for the new feature/change?
- [ ] Are unit tests added where appropriate?

**Config/Logging/Testability**
- [ ] Are all needed new/changed options added to default YAML files?
- [ ] Did we add proper logging messages for operator actions?
- [ ] Did we ensure compatibility with the previous version or cluster upgrade process?
- [ ] Does the change support oldest and newest supported PS version?
- [ ] Does the change support oldest and newest supported Kubernetes version?

[K8SPS-307]: https://perconadev.atlassian.net/browse/K8SPS-307?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: pkg/controller/ps/upgrade.go [hunk 1]
```diff
@@ -106,13 +106,46 @@ func (r *PerconaServerMySQLReconciler) smartUpdate(ctx context.Context, sts *app
...
+		}
+	}
+
+	retriable := func(err error) bool {
+		return err != nil
+	}
+
+	retry := k8sretry.DefaultRetry
+	retry.Duration = 5 * time.Second
+	retry.Steps = 10
```
Comment (egegunes) [inline pkg/controller/ps/upgrade.go:125]: please don't mutate this, you can create a custom `wait.Backoff` object

File: pkg/controller/ps/upgrade.go [hunk 2]
```diff
@@ -106,13 +106,46 @@ func (r *PerconaServerMySQLReconciler) smartUpdate(ctx context.Context, sts *app
...
+
+		if err != nil {
+			log.Info("primary pod deletion error", "pod", primPod.Name)
+			return err
+		}
+	}
+
+	retriable := func(err error) bool {
+		return err != nil
+	}
```
Comment (egegunes) [inline pkg/controller/ps/upgrade.go:121]: I think you can move this to inline in `k8sretry.OnError`

File: pkg/controller/ps/upgrade.go [hunk 3]
```diff
@@ -106,13 +106,46 @@ func (r *PerconaServerMySQLReconciler) smartUpdate(ctx context.Context, sts *app
...
+
+	retry := k8sretry.DefaultRetry
+	retry.Duration = 5 * time.Second
+	retry.Steps = 10
+
+	err = k8sretry.OnError(retry, retriable, func() error {
+
```
Comment (egegunes) [inline pkg/controller/ps/upgrade.go:128]: unnecessary emtpy line

File: pkg/controller/ps/upgrade.go [hunk 4]
```diff
@@ -106,13 +106,46 @@ func (r *PerconaServerMySQLReconciler) smartUpdate(ctx context.Context, sts *app
...
+		if primPod.ObjectMeta.Labels[controllerRevisionHash] != sts.Status.UpdateRevision {
+			return errors.New("primary pod controllerRevisionHash not equal sts.Status.UpdateRevision")
+		}
+		return nil
+	})
+
+	if err != nil {
+		log.Info("smart update of  primary pod did not finish correctly after 10 retries")
 		return nil
```
Comment (egegunes) [inline pkg/controller/ps/upgrade.go:142]: should we return nil in this case?

**Review by egegunes — CHANGES_REQUESTED** [review]

File: pkg/controller/ps/upgrade.go (hunk 1 above)
Comment (nmarukovich) [reply pkg/controller/ps/upgrade.go:125]: should we update it for deletePod method too?

**Review by hors — CHANGES_REQUESTED** [review]
@nmarukovich please add descriptions

**Comment by nmarukovich** [comment]
> @nmarukovich please add descriptions

fixed!

File: pkg/controller/ps/upgrade.go [hunk 5]
```diff
@@ -106,13 +107,42 @@ func (r *PerconaServerMySQLReconciler) smartUpdate(ctx context.Context, sts *app
...
+		}
+
+		if primPod.ObjectMeta.Labels[controllerRevisionHash] != sts.Status.UpdateRevision {
+			return errors.New("primary pod controllerRevisionHash not equal sts.Status.UpdateRevision")
+		}
 		return nil
+	})
+
+	if err != nil {
+		log.Info("smart update of  primary pod did not finish correctly after 10 retries")
```
Comment (egegunes) [inline pkg/controller/ps/upgrade.go:138]: we're doing 5 retries, aren't we?

File: pkg/controller/ps/upgrade.go [hunk 6]
```diff
@@ -106,13 +107,42 @@ func (r *PerconaServerMySQLReconciler) smartUpdate(ctx context.Context, sts *app
...
+		return err
 	}
 
-	return deletePod(ctx, r.Client, primPod, currentSet)
+	log.Info("primary pod updated", "primPod name", primPod.Name)
```
Comment (egegunes) [inline pkg/controller/ps/upgrade.go:142]:
```suggestion
	log.Info("primary pod updated", "pod", primPod.Name)
```

**Review by egegunes — CHANGES_REQUESTED** [review]

File: pkg/controller/ps/upgrade.go (hunk 5 above)
Comment (nmarukovich) [reply pkg/controller/ps/upgrade.go]: yes, sorry, missed it. fixed

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| async-ignore-annotations | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/async-ignore-annotations.log) |
| auto-config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/auto-config.log) |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/config.log) |
| config-router | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/config-router.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/demand-backup.log) |
| gr-demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-demand-backup.log) |
| gr-demand-backup-haproxy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-demand-backup-haproxy.log) |
| gr-finalizer | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-finalizer.log) |
| gr-haproxy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-haproxy.log) |
| gr-ignore-annotations | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-ignore-annotations.log) |
| gr-init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-init-deploy.log) |
| gr-one-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-one-pod.log) |
| gr-recreate | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-recreate.log) |
| gr-scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-scaling.log) |
| gr-security-context | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-security-context.log) |
| gr-self-healing | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-self-healing.log) |
| gr-tls-cert-manager | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-tls-cert-manager.log) |
| gr-users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-users.log) |
| haproxy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/haproxy.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/init-deploy.log) |
| limits | [failure](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/limits.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/monitoring.log) |
| one-pod | [failure](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/one-pod.log) |
| operator-self-healing | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/operator-self-healing.log) |
| recreate | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/recreate.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/scaling.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/service-per-pod.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/sidecars.log) |
| smart-update | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/smart-update.log) |
| tls-cert-manager | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/tls-cert-manager.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/users.log) |
| version-service | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/version-service.log) |
| We run 32 out of 32|

commit: https://github.com/percona/percona-server-mysql-operator/pull/677/commits/50b4e05fd12bbc8fa63e291ced49668bdeedf278
image: `perconalab/percona-server-mysql-operator:PR-677-50b4e05f`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| async-ignore-annotations | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/async-ignore-annotations.log) |
| auto-config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/auto-config.log) |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/config.log) |
| config-router | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/config-router.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/demand-backup.log) |
| gr-demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-demand-backup.log) |
| gr-demand-backup-haproxy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-demand-backup-haproxy.log) |
| gr-finalizer | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-finalizer.log) |
| gr-haproxy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-haproxy.log) |
| gr-ignore-annotations | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-ignore-annotations.log) |
| gr-init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-init-deploy.log) |
| gr-one-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-one-pod.log) |
| gr-recreate | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-recreate.log) |
| gr-scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-scaling.log) |
| gr-security-context | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-security-context.log) |
| gr-self-healing | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-self-healing.log) |
| gr-tls-cert-manager | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-tls-cert-manager.log) |
| gr-users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/gr-users.log) |
| haproxy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/haproxy.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/init-deploy.log) |
| limits | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/limits.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/monitoring.log) |
| one-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/one-pod.log) |
| operator-self-healing | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/operator-self-healing.log) |
| recreate | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/recreate.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/scaling.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/service-per-pod.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/sidecars.log) |
| smart-update | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/smart-update.log) |
| tls-cert-manager | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/tls-cert-manager.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/users.log) |
| version-service | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-677/50b4e05f/version-service.log) |
| We run 32 out of 32|

commit: https://github.com/percona/percona-server-mysql-operator/pull/677/commits/50b4e05fd12bbc8fa63e291ced49668bdeedf278
image: `perconalab/percona-server-mysql-operator:PR-677-50b4e05f`

**Review by egegunes — APPROVED** [review]

**Review by inelpandzic — APPROVED** [review]

**Review by hors — APPROVED** [review]

**Pull request merged**